packages/core/admin/admin/src/content-manager/components/RelationInput/components/RelationItem.js

```diff
@@ -1,52 +1,154 @@
+
+import { composeRefs } from '../../../utils';
 
 const ChildrenWrapper = styled(Flex)`
   width: 100%;
   /* Used to prevent endAction to be pushed out of container */
   min-width: 0;
 `;
 
-export const RelationItem = ({ children, disabled, endAction, style, ...props }) => {
+const RELATION_ITEM_DRAG_TYPE = 'RelationItem';
+
+export const RelationItem = ({
+  children,
+  canDrag,
+  disabled,
+  endAction,
+  style,
+  id,
+  index,
+  updatePositionOfRelation,
+  ...props
+}) => {
+  const relationRef = useRef(null);
+
+  const [{ handlerId }, dropRef] = useDrop({
+    accept: RELATION_ITEM_DRAG_TYPE,
+    collect(monitor) {
+      return {
+        handlerId: monitor.getHandlerId(),
+      };
+    },
+    hover(item, monitor) {
+      if (!relationRef.current) {
+        return;
+      }
+      const dragIndex = item.index;
+      const currentIndex = index;
+
+      // Don't replace items with themselves
+      if (dragIndex === currentIndex) {
+        return;
+      }
+
+      const hoverBoundingRect = relationRef.current.getBoundingClientRect();
+      const hoverMiddleY = (hoverBoundingRect.bottom - hoverBoundingRect.top) / 2;
+      const clientOffset = monitor.getClientOffset();
+      const hoverClientY = clientOffset.y - hoverBoundingRect.top;
+
+      // Dragging downwards
+      if (dragIndex < currentIndex && hoverClientY < hoverMiddleY) {
+        return;
+      }
+
+      // Dragging upwards
+      if (dragIndex > currentIndex && hoverClientY > hoverMiddleY) {
+        return;
+      }
+
+      // Time to actually perform the action
+      if (updatePositionOfRelation) {
+        updatePositionOfRelation(dragIndex, currentIndex);
+        item.index = currentIndex;
+      }
+    },
+  });
+
+  const [{ isDragging }, dragRef, dragPreviewRef] = useDrag({
+    type: RELATION_ITEM_DRAG_TYPE,
+    item: { index },
+    canDrag: canDrag && !disabled,
+    collect: (monitor) => ({
+      isDragging: monitor.isDragging(),
+    }),
+  });
+
+  const composedRefs = composeRefs(relationRef, dragRef);
+
   return (
-    <Box style={style} as="li">
-      <Flex
-        paddingTop={2}
-        paddingBottom={2}
-        paddingLeft={4}
-        paddingRight={4}
-        hasRadius
-        borderSize={1}
-        background={disabled ? 'neutral150' : 'neutral0'}
-        borderColor="neutral200"
-        justifyContent="space-between"
-        {...props}
-      >
-        <ChildrenWrapper justifyContent="space-between">{children}</ChildrenWrapper>
-        {endAction && <Box paddingLeft={4}>{endAction}</Box>}
-      </Flex>
+    <Box style={style} as="li" ref={dropRef}>
+      {isDragging ? (
+        <Box
+          ref={dragPreviewRef}
+          paddingTop={2}
+          paddingBottom={2}
+          paddingLeft={4}
+          paddingRight={4}
+          hasRadius
+          borderStyle="dashed"
+          borderColor="primary600"
+          borderWidth="1px"
+          background="primary100"
+          height="100%"
+        />
```

Comment:
I suppose this is the placeholder component? What do you think about defining this as a new component (in the same file) to improve the code readability a bit?

Reply:
Could do? we pass a lot of the styling props to the component, normally I would do `const Placeholder = styled(Box)` to make it clearer...

Let's make it a separate func with `forwardRef`.